File: Easy/0733-flood-fill/0733-flood-fill.java
class Solution {
    
    
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {                
        int original = image[sr][sc];
        
        if(original == color){
            return image;
        }
        
        int[] dx = {-1, 0, 1, 0};
        int[] dy = {0, 1, 0, -1};
        boolean[][] isVisited = new boolean[image.length][image[0].length];
        
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{sr, sc});
        
        
        image[sr][sc] = color;
        isVisited[sr][sc] = true;
        
        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            
            for(int i = 0; i < 4; i++){
                int nx = cur[0] + dx[i];
                int ny = cur[1] + dy[i];
                
                if(nx >= 0 && nx < image.length && ny >= 0 && ny < image[0].length && !isVisited[nx][ny]){
                    isVisited[nx][ny] = true;
                    if(image[nx][ny] == original){
                        image[nx][ny] = color;
                        queue.offer(new int[]{nx, ny});
                        
                    }                    
                }
                
            }
                        
        }
        
        return image;
        
    }
}

File: Easy/0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    const original = image[sr][sc];
    if(original === color){
        return image;
    }
    
    image[sr][sc] = color;
    const queue = [[sr, sc]];
    const dx = [-1, 0, 1, 0];
    const dy = [0, -1, 0, 1];
    
    while(queue.length > 0){
        const cur = queue.shift();
        let x = cur[0];
        let y = cur[1];
        
        for(let i = 0; i < 4; i++){
            let nx = x + dx[i];
            let ny = y + dy[i];
            
            if(nx >= 0 && nx < image.length && ny >= 0 && ny < image[0].length && image[nx][ny] == original){
                image[nx][ny] = color;
                queue.push([nx, ny]);
            }
        }
        
    }
    return image;
    
};